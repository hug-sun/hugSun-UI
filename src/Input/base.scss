@function getHeight($height){
  @return $height *2;
}   