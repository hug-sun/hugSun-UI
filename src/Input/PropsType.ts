import React, { ReactNode } from 'react';

/**
 * Input properties
 */
export default interface InputPropsType {
  accept?: string;
  autocomplete?: 'on' | 'off';
  autofocus?: boolean;
  max?: number;
  maxlength?: number;
  min?: number;
  multiple?: boolean;
  name?: string;
  placeholder?: string;
  readonly?: boolean;
  required?: boolean;
  type?:
    | 'button'
    | 'checkbox'
    | 'text'
    | 'textarea'
    | 'password'
    | 'reset'
    | 'submit'
    | 'number';
  value?: string;
  className?: string;
  prefixCls?: string;
  bordered?: boolean;
  id?: string;
  disabled?: boolean;
  size?: 'default' | 'large' | 'small';
  addonBefore: ReactNode;
  addonAfter: ReactNode;
  defaultValue?: string;
  [propName: string]: any;
}

export interface InputFocusOptions {
  cursor?: 'start' | 'end' | 'all';
}

export const defaultProps = {
  precls: 'h-input',
  bordered: true,
  size: 'default',
};
