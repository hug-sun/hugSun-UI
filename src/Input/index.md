---
title: Input
nav:
  title: 组件
  path: /components
---

## 基础使用

### type

```tsx
import React from 'react';
import { Input } from 'hugSun-UI';
const [val, setVal] = React.useState(null);
const handleChange = () => {
  console.log('change', 111);
};
const handlePressEnter = () => {
  console.log('PressEnter', 111);
};
const handleFocus = () => {
  console.log('focus');
};

const setProps = {
  defaultValue: 'test',
  className: 'test',
  size: 'large',
};
export default () => {
  return (
    <>
      <Input
        {...setProps}
        onFocus={handleFocus}
        //  maxlength={5}
      ></Input>
    </>
  );
};
```

<API></API>
