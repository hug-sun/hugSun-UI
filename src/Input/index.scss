.h-input-wrap{
  font-size: 14px;
  position: relative;
  display: flex;
  flex-wrap: nowrap; // 不换行
  width: 100%;
  justify-content: center;
  border: 1px solid #ddd;
 
  .h-input-addon{
    position: relative;
    padding: 0 11px;
    font-weight: normal;
    text-align: center;
    background-color: #fafafa;
    border-radius: 2px;
    transition: all 0.3s;
    max-width: 50px;
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    /* 设置内容垂直居中*/
    &::before{
      display: inline-block;
      content: "";
      height: 100%;
      vertical-align: middle;
    }
    /* --*/

    &:first-child {
      border-right: 1px solid #ddd;
    }
    &:last-child {
      border-left: 1px solid #ddd;
    }
  }
  // affix
  .h-input-affix-wrap {
    display: inline-flex;
    width: 100%;
    padding: 4px;  
    border-radius: 2px;
    // border: 1px solid #999;
    box-sizing: border-box;
    background-color: #fff;
    .h-input-affix {
      &::before{
        display: inline-block;
        content: "";
        height: 100%;
        vertical-align: middle;
      }
      padding:4px;
      .h-affix {
        width: 100%; 
      }
    }
  }
  .h-input {
    padding-left: 10px;
    border: none;
    flex:1;
    width: 100%;
    outline: none;
    height: 30px;
    line-height: 30px;
    &:focus {
      outline: none;
    }
    &.h-input-lg{
      height: 40px;
      line-height: 40px;
    }
    &.h-input-sm{
      height: 20px;
      line-height: 20px;
    }
    &.h-input-borderless {
      border: none
    }
    &:not(:first-child):not(:last-child) {
      border-radius: 0;
    }
  }
}









