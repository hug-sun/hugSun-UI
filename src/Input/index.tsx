import React, { useState, useRef, useEffect, ReactElement } from 'react';
import './index.scss';
import classnames from 'classnames';
import InputPropsType, { defaultProps, InputFocusOptions } from './PropsType';

const Input: React.FC<InputPropsType> = props => {
  const inputEl = useRef<HTMLInputElement>(null);
  const [oninput, setOninput] = useState<string | number>();
  // 设置input框的初始值
  const [value, setValue] = useState<string | number>();

  // type为number类型时候字符长度设置
  const setMaxLength = (event: any) => {
    let { maxlength, type } = props;
    if (maxlength && type == 'number') {
      let value = event.target.value;
      if (value && value?.length > maxlength) {
        setOninput(value?.slice(0, maxlength));
        return value?.slice(0, maxlength);
      }
      return;
    }
  };

  // 方法设置
  const triggerFocus = (
    element?: HTMLInputElement | HTMLTextAreaElement,
    option?: InputFocusOptions,
  ) => {
    if (!element) return;
    // element.focus(option);
  };

  // onchange事件后的处理
  const resolveOnChange = (
    target: HTMLInputElement | HTMLTextAreaElement | null, // input or inputArea
    e:
      | React.ChangeEvent<HTMLTextAreaElement | HTMLInputElement>
      | React.MouseEvent<HTMLElement, MouseEvent>
      | React.CompositionEvent<HTMLElement>,
    onChange:
      | undefined
      | ((
          event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
        ) => void),
    defaultValue: string | undefined,
  ) => {
    if (!onChange) {
      return;
    }

    if (!target) {
      return;
    }

    if (defaultValue) {
    }
    let event = e;
    console.log(111);
    // 事件类型判断并处理
    onChange(
      event as React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
    );
  };
  // onChange事件 输入框内容变化时候的回调
  const handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const { defaultValue, onChange } = props;
    setValue(e.target.value);
    resolveOnChange(inputEl.current, e, onChange, defaultValue);
  };

  //键盘事件处理
  const hanleKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
    const { onPressEnter, onKeyDown } = props;
    // enter 键
    if (onPressEnter && e.keyCode === 13) {
      onPressEnter(e);
      return;
    }
    onKeyDown?.(e);
  };

  // 聚焦事件
  const handleFocus = (e: React.FocusEvent<HTMLInputElement>) => {
    const { onFocus } = props;
    if (onFocus) {
      inputEl.current?.focus();
      onFocus(e);
    }
  };

  // 失焦事件
  const handleBlur = (e: React.FocusEvent<HTMLInputElement>) => {
    const { onBlur } = props;
    if (onBlur) {
      onBlur(e);
    }
  };
  // 渲染是否带有prefix的组件
  // const renderPrefix = () => {
  //   const { prefix, suffix } = props;

  //   const _prefix = prefix ? (
  //     <span className={styles.prefix} onClick={this.emptyValue}>{prefix}</span>
  //   ) : null;

  //   const closeBtn = (<i className={styles.closeBtn} onClick={this.emptyValue}></i>)

  //   const _suffix = this.state.isHasCloseBtn ? closeBtn : (
  //     <span className={styles.suffix}>
  //       {
  //         suffix ? suffix : null
  //       }
  //     </span>
  //   )

  //   return (
  //     <span className={styles.inputGroupWrapper}>
  //       {_prefix}
  //       {React.cloneElement(children)}
  //       {_suffix}
  //     </span>
  //   );
  // }

  // 渲染是否带有addon的组件
  const RenderAddonInput = (children: ReactElement) => {
    const { addonBefore, addonAfter } = props;
    // if (!addonAfter || !addonBefore) return;
    const _addonBefore = addonBefore ? (
      <span className="h-input-addon">{addonBefore}</span>
    ) : null;

    const _addonAfter = addonAfter ? (
      <span className="h-input-addon">{addonAfter}</span>
    ) : null;

    return (
      <span className="h-input-wrap">
        {_addonBefore}
        {React.cloneElement(children)}
        {_addonAfter}
      </span>
    );
  };

  // 渲染input
  const renderInput = () => {
    const {
      className,
      id,
      disabled,
      bordered,
      size,
      type,
      placeholder,
      defaultValue,
    } = props;

    const otherProps: any = {};
    const otherPorpsArray: Array<string> = ['maxLength'];

    // class设置
    const setClassName = () => {
      let { precls } = defaultProps;
      let cls: string = precls;
      return classnames(cls, className, {
        [`${precls}-sm`]: size === 'small',
        [`${precls}-lg`]: size === 'large',
        [`${precls}-disabled`]: disabled,
        [`${precls}-borderless`]: bordered,
      });
    };

    return (
      <>
        <input
          ref={inputEl}
          placeholder={placeholder ? placeholder : '请输入内容'}
          className={setClassName()}
          disabled={disabled}
          id={id}
          type={type ? type : 'text'}
          onKeyDown={hanleKeyDown}
          onFocus={handleFocus}
          onInput={setMaxLength}
          defaultValue={defaultValue}
          onChange={handleOnChange}
          value={oninput}
        />
      </>
    );
  };

  return <>{RenderAddonInput(renderInput())}</>;
};
export default Input;
