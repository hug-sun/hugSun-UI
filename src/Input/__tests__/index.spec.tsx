import React from 'react';
import { shallow } from 'enzyme';
import Input from '..';
import toJson from 'enzyme-to-json';

describe('Input', () => {
  it('Input render correctly', () => {
    const wrapper = shallow(<Input>default</Input>);

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
