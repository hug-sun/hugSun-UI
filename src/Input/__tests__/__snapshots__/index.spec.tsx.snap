// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Input render correctly 1`] = `
<div>
  <input
    placeholder="hello"
  />
</div>
`;
